### PR TITLE
fix(quickstart): preserve schema-defined channel fields

### PR DESCRIPTION
**Pull request by Audacity88**
## Summary

- **Base branch:** `master`
- **What changed and why:**
  - Replace the quickstart channel submission's single optional token with a schema-keyed field map, matching the existing model-provider field contract.
  - Have the CLI store each prompted `FieldDescriptor.key`, have zerocode preserve each rendered row key, and have the web dashboard submit its staged descriptor-keyed field map directly.
  - Validate submitted channel keys against the fields advertised for that channel type and reject unadvertised keys before config writes.
  - Keep Telegram token requirements in the config schema: enabled aliases require a usable `bot_token`, and the `<unset>` display sentinel is never accepted as persisted config.
  - Add regression coverage for serialization, Telegram validation, rejected field keys, encrypted persistence, multi-channel binding, zerocode wire shape, and generic field collection.
- **Scope boundary:** This does not change channel networking, polling/webhook lifecycle, peer-group routing, or message delivery. It does change config validation for enabled Telegram aliases with blank or unset tokens. It does not migrate already-broken config files automatically or cover the separate second-agent follow-up reported in #8505.
- **Blast radius:** Quickstart channel creation in the CLI, zerocode, and web dashboard; quickstart's channel wire shape; config validation for Telegram aliases; and CLI localization catalogs.
- **Linked issues:** Related #8505. Broader first-run system/E2E coverage is tracked in #8766.
- **Labels:** `bug`, `quickstart`, `core`, `config`, `runtime`, `zerocode`, `risk:high`, `size:L`

## Validation Evidence (required)

```bash
cargo +1.96.1 fmt --all -- --check
git diff --check
cargo +1.96.1 clippy --workspace --exclude zeroclaw-desktop --all-targets --features ci-all -- -D warnings
cargo test -p zeroclaw-config channel_quickstart --lib
cargo test -p zeroclaw-config telegram --lib
cargo test -p zeroclaw-runtime channel_fields --lib
cargo test -p zeroclaw-runtime multiple_channels_all_bind_to_agent --lib
cargo test -p zerocode channel_ --bin zerocode
cargo test --bin zeroclaw --features agent-runtime --no-run
cargo web build
```

- Config quickstart tests: 2 passed.
- Config Telegram tests: 19 passed.
- Runtime channel-field tests: 5 passed.
- Runtime multi-channel regression: 1 passed.
- Zerocode channel tests: 4 passed.
- Agent-runtime binary compiled successfully.
- The web build generated the OpenAPI client and completed TypeScript and Vite production builds; 2,164 modules were transformed.
- Rust 1.96.1 formatting, full workspace Clippy, and whitespace checks passed.
- The linker emitted the existing macOS compact-unwind size warning for the runtime test binary; tests still linked and passed.
- Vite emitted non-blocking chunk-size warnings for existing large bundles.

- **Beyond CI:** Verified persisted Telegram and Discord test credentials decrypt to the submitted placeholder values, and unknown or unadvertised keys are rejected before mutating config.
- **Live Telegram smoke:** On the current head with matching binaries and a clean daemon socket, entered a real bot token through CLI Quickstart, confirmed `channels.telegram.<alias>.bot_token` reported `populated: true` after reload without printing the value, ran `channel doctor` successfully, started the listener, received an inbound `/start`, and observed the outbound bot response in Telegram Desktop.
- **ZeroCode persistence smoke:** On the current head with the TUI and daemon from the same build, entered the same credential through ZeroCode Quickstart, completed apply/reload, and confirmed the canonical `bot_token` path reported `populated: true` from a separate config read without printing the value.
- **Redaction check:** The credential remained behind secret-field/config redaction, and a token-shape scan found no Telegram credential in the runtime trace.
- **Not covered by this smoke:** Creating or reusing a second Telegram agent; the reporter's follow-up remains recorded in #8505.

## Security & Privacy Impact (required)

- New permissions, capabilities, or file system access scope? `No`
- New external network calls? `No`
- Secrets / tokens / credentials handling changed? `Yes`
- PII, real identities, or personal data in diff, tests, fixtures, or docs? `No`
- If any `Yes`, describe the risk and mitigation: Quickstart now carries schema-keyed channel fields, including secret fields, through a generic map. The runtime rejects unadvertised fields before writing config, and the unknown-key regression verifies that its error does not expose the submitted secret value. Accepted secret fields persist through the existing encrypted config path. Tests use placeholder token-shaped values only.

## Compatibility (required)

- Backward compatible? `No` for clients that construct the quickstart submission JSON directly; bundled clients are updated together.
- Config / env / CLI surface changed? `Yes`
- Details: `ChannelQuickStart.token` is replaced by `ChannelQuickStart.fields`. Zerocode's existing exact package-version handshake rejects daemon/TUI version mismatches. The persisted config format is unchanged. Enabled Telegram aliases with blank tokens now fail validation, and `<unset>` is rejected for both enabled and disabled aliases.

## Rollback (required for medium/high-risk PRs)

- **Fast rollback command/path:** `git revert <merge-commit-sha>`
- **Feature flags or config toggles:** None.
- **Observable failure symptoms:** A bundled quickstart surface cannot submit an advertised channel field, a failed fresh channel leaves partial config behind, or valid Telegram quickstart input is rejected before channel startup.

## Supersede Attribution (required only when `Supersedes #` is used)

- Superseded PRs + authors: N/A
- Scope materially carried forward: N/A
- `Co-authored-by` trailers added in commit messages for incorporated contributors? `No`
- If `No`, why: No superseded contribution is incorporated.

### CODE REVIEW

**Review by IftekharUddin — CHANGES_REQUESTED** [review]
Changes requested. Preserving the schema’s canonical `bot_token` fixes the reported Telegram submission loss, and the enabled-channel validation is useful. The token-key mapping is now duplicated across CLI and ZeroCode, however, and the copies already have different behavior.

### 🔴 Blocking — Define token-field resolution once and reuse it on both surfaces

`src/main.rs::quickstart_channel_token_key` accepts `bot_token`, `bot-token`, `token`, and `access-token`; `apps/zerocode::quickstart_channel_token_key` omits `access-token`. These functions encode the same quickstart submission fact in two places, directly violating the repository’s single-source-of-truth rule and reproducing the drift that caused this bug.

The CLI implementation has another edge case: it consumes a `BTreeMap` and returns the first matching key without trimming or skipping empty values. Because aliases sort before `bot_token`, an empty legacy alias can mask a populated canonical value.

```mermaid
flowchart LR
    S["Schema field descriptors\ncanonical keys"] --> R["one shared token resolver"]
    R --> C["CLI quickstart"]
    R --> Z["ZeroCode quickstart"]
```

Move one resolver into the shared quickstart/runtime surface and call it from both frontends. Give the canonical schema key deterministic precedence, retain only intentionally supported legacy aliases, trim values, and skip empty/`UNSET` inputs. Ideally consume the schema descriptor’s secret/canonical metadata rather than maintaining a generic token-name list at all.

Add table-driven parity tests for every accepted key, canonical-plus-legacy conflicts, empty aliases, whitespace, and unset display values on both callers.

### 🟡 Warning — Complete live secret-path validation before readiness

The focused tests and full CI are green, but the PR body correctly leaves the real Telegram path unverified. Before marking this high-risk secret-handling draft ready, run the requested disposable bot smoke: enter a token through both CLI and ZeroCode, verify the persisted canonical field without printing the value, reload, run channel doctor, and receive/send one Telegram message. Include a redaction check for logs and UI state.

### 🟢 What looks good — Missing tokens now fail early and locally

- Enabled Telegram channels now fail validation when the canonical token is truly absent.
- Disabled/staged Telegram entries remain valid.
- The new CLI error has entries in every supported runtime locale.
- Tests use placeholder token-shaped values and do not print a real credential.

Centralizing the resolver is important here: this patch should remove the class of cross-surface drift, not add another pair of alias tables that must be kept synchronized.

File: apps/zerocode/src/quickstart_pane.rs [hunk 1]
```diff
@@ -689,6 +689,22 @@ struct FieldFormRow {
     buf: String,
 }
 
+fn quickstart_channel_token_key(key: &str) -> bool {
```
Comment (IftekharUddin) [inline apps/zerocode/src/quickstart_pane.rs]: ### 🔴 Blocking — Reuse one canonical token-field resolver

This alias table is duplicated in `src/main.rs`, and the copies already differ: CLI accepts `access-token`, while ZeroCode does not. That is the same cross-surface drift class behind this bug and violates `AGENTS.md`’s single-source-of-truth rule. Put one resolver in the shared quickstart/runtime layer (preferably driven by schema descriptor metadata) and call it from both surfaces.

File: src/main.rs [hunk 2]
```diff
@@ -1216,6 +1216,23 @@ fn apply_homebrew_onboard_config_dir() {
     );
 }
 
+#[cfg(feature = "agent-runtime")]
+fn quickstart_channel_token_key(key: &str) -> bool {
```
Comment (IftekharUddin) [inline src/main.rs]: ### 🔴 Blocking — Make precedence and empty-value handling canonical

This second alias table duplicates the ZeroCode implementation. Because `extras` is a `BTreeMap`, `find_map` selects aliases lexicographically and does not skip blank values, so an empty `bot-token`/`access-token` can mask a populated `bot_token`. Use the single shared resolver, prefer the canonical schema key deterministically, trim values, and ignore empty/unset aliases; cover conflicts in parity tests.

**Comment by Audacity88** [comment]
@IftekharUddin, the current head addresses the resolver blocker by removing token-alias resolution from both frontends. CLI and ZeroCode now round-trip the daemon's schema descriptor keys, while the runtime validates advertised keys and writes them through the canonical config paths.

The requested live validation is complete:

- On the current head with matching binaries and a clean daemon socket, CLI Quickstart persisted the canonical Telegram `bot_token`; a separate config read after reload reported `populated: true` without printing the value.
- `channel doctor` was healthy, the listener received an inbound `/start`, and the bot sent a visible response in Telegram Desktop.
- ZeroCode Quickstart, using the TUI and daemon from the same current-head build, also persisted the canonical `bot_token`; a separate config read after apply/reload reported `populated: true` without printing the value.
- A token-shape scan found no Telegram credential in the runtime trace.

This smoke did not cover creating or reusing a second Telegram agent; the reporter's separate follow-up remains recorded in #8505.

All required CI checks are green on the current head. Could you re-review when convenient?

**Review by IftekharUddin — APPROVED** [review]
@Audacity88, I re-reviewed current head 11e35f3b44ef11d680949dc1e04f2a99b1f308f3 against my request-changes review at 7e0b145273efcc2dbe596e9acc7f6d48c78ec227, both blocking inline threads, the complete thirteen-file diff, the CLI/ZeroCode/web submission paths, runtime field validation and atomic config writes, Telegram schema validation, the live smoke report, #8505 and #8766, current metadata, and exact-head required CI. I did not rerun the local Rust/web commands or the credentialed Telegram smoke.

### ✅ Resolved — schema descriptor keys are the single source of truth

The duplicated CLI and ZeroCode token-alias resolvers are gone. All three bundled frontends now retain the daemon-advertised `FieldDescriptor.key` and submit a generic field map, while the runtime derives its allowlist from `field_shape`, rejects unadvertised keys, trims and drops blank/`<unset>` values, and persists accepted values through the canonical config path. That removes the divergent alias tables rather than centralizing another hand-maintained copy, fully resolving both inline blockers.

The runtime stages each fresh channel against a cloned `Config` and only commits after every field, `enabled`, and Telegram token check succeeds. The unknown/unadvertised-key regressions prove rejected input neither mutates config nor leaks the submitted secret, and the persistence tests prove canonical Telegram and Discord credentials still use the encrypted secret path. `TelegramConfig::validate_bot_token` is now the canonical enabled/placeholder rule used by both quickstart and full config validation.

### ✅ Resolved — the requested live credential path is covered

The current-head report records a disposable Telegram flow through CLI Quickstart: canonical `bot_token` persistence checked without printing the value, reload, healthy channel doctor, listener startup, inbound `/start`, and visible outbound response. It separately records ZeroCode apply/reload persistence and a runtime-trace token-shape scan. The untested second-agent case remains explicitly bounded to #8505, with broader first-run E2E ownership in #8766. That satisfies the prior readiness warning without broadening this PR.

### 🟡 Warning — refresh the high-risk merge result before landing

The branch is 24 commits behind `master`, and intervening base changes touch `crates/zeroclaw-config/src/presets.rs`, `schema.rs`, `traits.rs`, and `src/main.rs`. GitHub still reports a clean merge and exact-head required CI is green, so I am not reopening the implementation review; because this is a high-risk config/secret path, refresh the branch and require current CI on the resulting merge tree before landing.

The public body also uses the superseded `Validation Evidence` layout. Before merge, please carry its unusually strong evidence into the current template and explicitly answer reviewer testing, CI checks relied on/coverage gap, prompt-injection impact, and Rust/MSRV floor. The existing text already supports the answers; this is a merge-record correction, not a code blocker.

### 🟢 Metadata and milestone

The live `bug`, `core`, `config`, `runtime`, `quickstart`, `zerocode`, `risk:high`, and `size:L` labels match the cross-surface wire/config change. `Operator UX: Onboarding, Pairing & Self-Service` is the correct milestone because its charter explicitly includes guided onboarding and quickstart credential preservation.

No blocking findings remain. I approve this head, subject to the refreshed merge-tree gate above.

**Review by WareWolf-MoonWall — APPROVED** [review]
### 🟢 What looks good — correct migration from single-token to schema-keyed field map

`ChannelQuickStart.token: Option<String>` → `fields: HashMap<String, String>` is the right shape: it matches the existing model-provider field contract and removes the brittle hardcoded `bot-token`/`token` fallback cascade. CLI stores each `FieldDescriptor.key`; ZeroCode preserves rendered row keys; web dashboard submits the descriptor-keyed field map directly. Submitted channel keys are validated against the fields advertised for that channel type; unadvertised keys are rejected before config writes.

### 🟢 What looks good — Telegram validation tightened in schema

Enabled Telegram aliases with blank or `<unset>` tokens now fail validation at the schema layer rather than producing a silently broken config. The `<unset>` display sentinel is correctly rejected for both enabled and disabled aliases.

### 🟢 What looks good — regression coverage

New tests cover: serialization round-trip, Telegram validation, rejected field keys, encrypted persistence, multi-channel binding, ZeroCode wire shape, and generic field collection. Live smoke confirmed real Telegram bot token through CLI Quickstart and ZeroCode Quickstart. CI 27/27 ✅. All five locale catalogs updated.

**Review by JordanTheJet — COMMENTED** [review]
Reviewed the full 13-file diff independently — agrees with the two existing approvals; no blockers from me.

**What looks good**
- The `token: Option<String>` → `fields: HashMap<String, String>` migration is the right shape: it matches the model-provider field contract, and having the runtime derive its allowlist from `field_shape` (rejecting unadvertised keys) removes the hardcoded fallback cascade that caused #8505 rather than adding another alias table to keep in sync.
- Staging each fresh channel against a cloned `Config` and only committing after every field, `enabled`, and the Telegram token check succeeds is the correct atomicity guarantee. The unknown-key regression convincingly shows rejected input neither mutates config nor leaks the submitted secret.
- `TelegramConfig::validate_bot_token` as the single enabled/placeholder rule (shared by quickstart and full config validation), with `<unset>` rejected for both enabled and disabled aliases, closes the silent-broken-config gap cleanly. Locale coverage across all five catalogs is a nice touch.

**Nits (non-blocking)**
- `usable_quickstart_value` is just `is_unset_display_value` negated — a one-line doc tying them together would help future readers.
- The `fields.sort_by_key` sort looks purely for determinism; a short comment would prevent it being read as a functional requirement.

**Before landing (not code changes)**
- The branch is ~24 commits behind `master`, and those commits touch the same config/secret files this PR edits (`presets.rs`, `schema.rs`, `traits.rs`, `main.rs`). For a high-risk secret path, please refresh the branch and require current CI on the resulting merge tree before merging.
- The body still uses the superseded `Validation Evidence` layout — worth carrying its (strong) evidence into the current template as a merge-record correction.

Recommending approve-with-nits, subject to the merge-tree refresh gate above.
